src/client/dfuse/ops/readdir.c

```diff
@@ -7,273 +7,437 @@
 
-	dfuse_compute_inode(udata->inode->ie_dfs, &oid, &stbuf.st_ino);
+	oh->doh_anchor_index += count;
+	oh->doh_dre_index = 0;
+	oh->doh_dre_last_index = count;
 
-	/*
-	 * If we are still within the fuse size limit (less than 4k - we have
-	 * not gone beyond 4k and cur_off is still 0).
-	 */
-	if (oh->doh_cur_off == 0) {
-		/** try to add the entry within the 4k size limit. */
-		ns = fuse_add_direntry(udata->req, oh->doh_buf + udata->b_off,
-				       udata->fuse_size - udata->b_off, name,
-				       &stbuf, oh->doh_fuse_off + 1);
-
-		/** if entry fits, increment the stream and fuse buf offset. */
-		if (ns <= udata->fuse_size - udata->b_off) {
-			udata->b_off += ns;
-			oh->doh_fuse_off++;
-			D_GOTO(out, rc = 0);
-		}
+	DFUSE_TRA_DEBUG(oh, "Added %d entries, anchor_index %d rc %d",
+			count, oh->doh_anchor_index, rc);
 
-		/*
-		 * If entry does not fit within the 4k fuse imposed size, we now
-		 * add the entry, but within the larger size limitation of the
-		 * OH buffer (16k). But we also need to save the state of the
-		 * current offset since this will not be returned in the current
-		 * readdir call but will be consumed in subsequent calls.
-		 */
-		oh->doh_start_off[oh->doh_idx] = udata->b_off;
-		oh->doh_cur_off = udata->b_off;
-		oh->doh_dir_off[oh->doh_idx] = oh->doh_fuse_off;
-
-		ns = fuse_add_direntry(udata->req, oh->doh_buf + udata->b_off,
-				       udata->size - udata->b_off, name, &stbuf,
-				       oh->doh_dir_off[oh->doh_idx] + 1);
-
-		/** Entry should fit now */
-		D_ASSERT(ns <= udata->size - udata->b_off);
-		oh->doh_cur_off += ns;
-		oh->doh_dir_off[oh->doh_idx]++;
-
-		/** no need to issue further dfs_iterate() calls. */
-		udata->stop = 1;
-		D_GOTO(out, rc = 0);
+	if (count) {
+		if (daos_anchor_is_eof(&oh->doh_anchor))
+			oh->doh_dre[count - 1].dre_next_offset = READDIR_EOD;
+	} else {
+		*eod = true;
 	}
 
-insert:
-	/*
-	 * At this point, we are already adding to the buffer within the large
-	 * size limitation where it will be consumed in future readdir calls.
-	 */
-	ns = fuse_add_direntry(udata->req, oh->doh_buf + oh->doh_cur_off,
-			       udata->size - oh->doh_cur_off, name, &stbuf,
-			       oh->doh_dir_off[oh->doh_idx] + 1);
-	/*
-	 * In the case where the OH handle does not fit, we still need to add
-	 * the entry since DFS already enumerated it. So, realloc to fit the
-	 * entries that were already enumerated and insert again.
-	 */
-	if (ns > udata->size - oh->doh_cur_off) {
-		udata->size = udata->size * 2;
-		oh->doh_buf = realloc(oh->doh_buf, udata->size);
-		if (oh->doh_buf == NULL)
-			D_GOTO(out, rc = -ENOMEM);
-		goto insert;
+	return rc;
+}
+
+static int
+create_entry(struct dfuse_projection_info *fs_handle,
+	     struct dfuse_inode_entry *parent,
+	     struct fuse_entry_param *entry,
+	     dfs_obj_t *obj,
+	     char *name,
+	     char *attr,
+	     daos_size_t attr_len,
+	     d_list_t **rlinkp)
+{
+	struct dfuse_inode_entry	*ie;
+	d_list_t			 *rlink;
```

Comment:
Only if you need to push again:
```suggestion
	d_list_t			*rlink;
```

Reply:
Will fix in a later PR.